core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -98,8 +98,15 @@ class BlockManagerMasterEndpoint(
 
     case _updateBlockInfo @
         UpdateBlockInfo(blockManagerId, blockId, storageLevel, deserializedSize, size) =>
-      context.reply(updateBlockInfo(blockManagerId, blockId, storageLevel, deserializedSize, size))
-      listenerBus.post(SparkListenerBlockUpdated(BlockUpdatedInfo(_updateBlockInfo)))
+      val succeed = updateBlockInfo(blockManagerId, blockId, storageLevel, deserializedSize, size)
+      context.reply(succeed)
+      // SPARK-30594: we should not post `SparkListenerBlockUpdated` when updateBlockInfo
+      // returns false, which may causes negative memory usage in UI(see SPARK-30465). In
+      // this case, that BlockManager will re-register and report all blocks to driver later.
+      // So, we should avoid to post the same block for two times.
+      if (succeed) {
```

Comment:
We should be able to add one test case for it in `BlockManagerSuite` ?

Reply:
Add UT in https://github.com/apache/spark/pull/27418/commits/f3fabdc812917c3e13283d380b99656ecc06cd93